python/whylogs/core/resolvers.py

```diff
@@ -16,11 +16,85 @@
 ColumnSchema: TypeAlias = "ColumnSchema"  # type: ignore
 
 
+@dataclass
+class MetricSpec:
+    """
+    Specify a Metric to instantiate.
+    """
+
+    metric: Any  # Should be a subclass of Metric, it should be the class, not an instance
+    config: Optional[MetricConfig] = None  # omit to use default MetricConfig
+
+
+@dataclass
+class ResolverSpec:
+    """
+    Specify the metrics to instantiate for matching columns. column_name
+    takes precedence over column_type. column_type should be a subclass
+    of DataType, i.e., AnyType, Frational, Integral, or String. Pass the
+    class, not an instance.
+    """
+
+    column_name: Optional[str] = None  # TODO: maybe make this a regex
+    column_type: Optional[Any] = None
+    metrics: List[MetricSpec] = field(default_factory=list)
+
+
 class Resolver(ABC):
     """A resolver maps from a column name and a data type to trackers.
 
     Note that the key of the result dictionaries defines the namespaces of the metrics in the serialized form."""
 
+    def __init__(self):
+        self.additional_specs: List[ResolverSpec] = []
+
+    def add_resolver_spec(self, resolver_spec: ResolverSpec):
+        """Add resolver specification to existing resolver instance.
+
+        Parameters
+        ----------
+        resolver_spec : ResolverSpec
+            Resolver specification that define the metrics to instantiate for matching columns.
+        """
+        if resolver_spec.column_name and resolver_spec.column_type:
+            logger.warning(
+                f"Resolver Spec: column {resolver_spec.column_name} also specified type, name takes precedence"
+            )
+        if not (resolver_spec.column_name or resolver_spec.column_type):
+            raise ValueError("Resolver Spec: resolver specification must supply name or type")
+        for metric_spec in resolver_spec.metrics:
+            if not issubclass(metric_spec.metric, Metric):
+                raise ValueError("Resolver Spec: must supply a Metric subclass to MetricSpec")
+        self.additional_specs.append(resolver_spec)
+
+    def resolve_additional_specs(
+        self, result: Dict[str, Metric], name: str, why_type: DataType, column_schema: ColumnSchema
+    ) -> Dict[str, Metric]:
+        """Resolves resolver specifications added by add_resolver_spec
+
+        Parameters
+        ----------
+        result : Dict[str, Metric]
+            Dictionary of metrics of the resolver before resolving additional specs.
+        name : str
+
+        Returns
+        -------
+        Dict[str, Metric]
+            Dictionary of metrics related to the additional resolver specificiations.
+        """
+        additional_result = {}
+        for resolver_spec in self.additional_specs:
+            col_name, col_type = resolver_spec.column_name, resolver_spec.column_type
+
+            if (col_name and col_name == name) or (col_name is None and isinstance(why_type, col_type)):  # type: ignore
```

Comment:
This looks very similar to the code in `DeclarativeResolver`, and doesn't respect the {en,dis}able members of the `MetricConfig`. Perhaps refactor to share code with `DeclarativeSchema`?  Also, maybe this should be an abstract method in the `Resolver` base class?

Reply:
Absolutely!

> Perhaps refactor to share code with DeclarativeSchema?

With latest changes, I'm now using DeclarativeResolver to resolve the additional metrics.

> Also, maybe this should be an abstract method in the Resolver base class?

I changed `resolve_additional_specs` to be abstract in Resolver Base Class, and implemented it in each existing subclass. Is that what you meant?